# **Titanic Survival Prediction:Machine Learning Fundamentals**

# **INTRODUCTION**

This notebook explores fundamental machine learning concepts using the titanic dataset.

The objective is to understand the machine learning workflow including:

-Supervised and unsupervised learning .

-Train test splitting.

-Model training

-model Evaluation

-Overfitting and Underfitting

The target variable is passenger survival:
0= Did not survive
1=Survived

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report


In [9]:
df=pd.read_csv("/content/drive/MyDrive/TITANIC CLEANED.csv")
df.head()

,PassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_Known,Embarked_Q,Embarked_S
0,1,0,3,0,22.0,1.0,0.0,7.2500,0,0,1
1,2,1,1,1,38.0,1.0,0.0,71.2833,1,0,0
2,3,1,3,1,26.0,0.0,0.0,7.9250,0,0,1
3,4,1,1,1,35.0,1.0,0.0,53.1000,1,0,1
4,5,0,3,0,35.0,0.0,0.0,8.0500,0,0,1


In [8]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


After running the cell above and following the prompts to authenticate, your Google Drive will be mounted at `/content/drive`.

You will then need to update the `pd.read_csv()` function in the next cell to include the full path to your file, for example:

```python
df = pd.read_csv('/content/drive/MyDrive/your_folder/TITANIC CLEANED.csv')
```

**Please replace `your_folder` with the actual path to your file on Google Drive, and make sure to include the `.csv` extension if your file has it.**

In [10]:
df=pd.read_csv("/content/drive/MyDrive/TITANIC CLEANED.csv")
df.head()

,PassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_Known,Embarked_Q,Embarked_S
0,1,0,3,0,22.0,1.0,0.0,7.2500,0,0,1
1,2,1,1,1,38.0,1.0,0.0,71.2833,1,0,0
2,3,1,3,1,26.0,0.0,0.0,7.9250,0,0,1
3,4,1,1,1,35.0,1.0,0.0,53.1000,1,0,1
4,5,0,3,0,35.0,0.0,0.0,8.0500,0,0,1


In [12]:
df.shape

(891, 11)

The Titanic dataset contains information about 891 passengers.
The objective is to use passenger characteristics to predict wether a passenger survived.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    int64  
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    float64
 6   Parch        891 non-null    float64
 7   Fare         891 non-null    float64
 8   Cabin_Known  891 non-null    int64  
 9   Embarked_Q   891 non-null    int64  
 10  Embarked_S   891 non-null    int64  
dtypes: float64(4), int64(7)
memory usage: 76.7 KB


# **FEATURE SELECTION**
The target variable is survived where:
-0: Did not survive
-1:Survive

The remaining variables are used as predictor variables(features)

In [13]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Cabin_Known', 'Embarked_Q', 'Embarked_S'],
      dtype='object')

In [14]:
x=df.drop("Survived",axis=1)
y=df["Survived"]

In [15]:
x.head()

,PassengerId,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_Known,Embarked_Q,Embarked_S
0,1,3,0,22.0,1.0,0.0,7.2500,0,0,1
1,2,1,1,38.0,1.0,0.0,71.2833,1,0,0
2,3,3,1,26.0,0.0,0.0,7.9250,0,0,1
3,4,1,1,35.0,1.0,0.0,53.1000,1,0,1
4,5,3,0,35.0,0.0,0.0,8.0500,0,0,1


In [16]:
y.head()

,Survived
0,0
1,1
2,1
3,1
4,0


x -> input features which is everything the model uses.

y-> target variable which is what the model predicts.

# **SUPERVISED AND UNSUPERVISED LEARNING**

## Supervised Learning

Supervised learning is a type of machine learning where a model learns from labelled data. This means that the dataset contains both input variables (features) and a known output variable (target).

The aim of supervised learning is to learn the relationship between the features and the target variable so that predictions can be made on new, unseen data.

In this Titanic dataset, supervised learning is applied because the target variable, **Survived**, is already known.

The model uses passenger characteristics such as:
- Passenger class
- Age
- Fare
- Other available passenger information

to learn patterns associated with survival outcomes.

Examples of supervised learning algorithms include:
- Logistic Regression
- Decision Trees
- Random Forest
- Support Vector Machines

Real-world applications of supervised learning include:
- Predicting customer credit risk
- Email spam classification
- Medical diagnosis prediction
- Fraud detection

## Unsupervised Learning

Unsupervised learning is a type of machine learning where a model learns patterns from data that does not have labelled outcomes.

The goal is to discover hidden structures, relationships, or groups within the dataset without being given a specific answer to predict.

Examples of unsupervised learning algorithms include:
- K-Means Clustering
- Hierarchical Clustering
- Principal Component Analysis (PCA)

Real-world applications of unsupervised learning include:
- Customer segmentation
- Grouping similar products
- Detecting unusual patterns in data

Unlike supervised learning, unsupervised learning does not predict a known target variable. Instead, it explores the underlying structure of the data.


# **TRAIN AND TEST SPLIT**

# Purpose of Splitting the Dataset

In ML, datasets are divided into training and testing set to evaluate how wella model can learn patterns and make predictions on new, unseen data.

#TRAINING SET

The training set is a portion of the dataset used to teach the ML model.The model analyzes the relationship between passenger featuresand survival outcomes to learn patterns that could be useful for prediction.

#TESTING SET

The testing set contains  data that the model has not seen during training.It is used to evaluate wether the model has learned meaningful patterns that can be applied to new observations.Using unseen testing data helps determine whether the model can generalize beyond the passenger it was trained on

In this project, the titanic dataset will be divided into:
- 80% training data
- 20% testing data

A random state is also going to be used to ensure that the same random split is produced each timethe analysis is repeated.





In [17]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)
#

# **OVERFITTING AND UNDERFITTING**

When building a machine learning model, the goal is not only to perform well on the data it was trained on, but also to make accurate predictions on new, unseen data.

# OVERFITTING

Overfitting occurs when a machine learning model learns the training data too closely, including noise and random patterns that do not represent the true relationship in the data.

As a result, the model performs very well on the training data but performs poorly on unseen testing data.

In the Titanic dataset, an overfitted model may memorize specific patterns from the training passengers instead of learning general relationships between passenger characteristics and survival outcomes.

For example, a model may learn very specific combinations of passenger features that only exist in the training data but fail to predict survival accurately for new passengers.

# EFFECTS OF OVERFITTING

- High accuracy on training data.
- Poor performance on testing data.
- Reduced ability to generalize to new observations.


# REDUCING OVERFITTING

- Use more training data.
- Reduce unnecessary features.
- Use simpler models.
- Apply regularization techniques.
- Use cross-validation to evaluate model performance.

## UNDERFITTING

Underfitting occurs when a machine learning model is too simple to capture important patterns in the data.

The model performs poorly on both the training data and testing data because it has not learned enough information from the dataset.

In the Titanic dataset, an underfitted model may fail to capture important relationships between passenger characteristics such as age, passenger class, fare, and survival.

#EFFECTS OF UNDERFITTING

- Poor performance on training data.
- Poor performance on testing data.
- The model cannot identify important patterns.

# REDUCING UNDERFITTING

- Use a more complex model.
- Include more relevant features.
- Improve feature selection.
- Train the model for longer where applicable.


A good machine learning model should find the balance between overfitting and underfitting.

The aim is to create a model that learns meaningful patterns from the Titanic dataset and can accurately predict survival outcomes for passengers it has not seen before.




## **Model Evaluation Basics**

## Evaluating Machine Learning Models

After a machine learning model has been trained, it is important to evaluate how well it performs. Model evaluation helps determine whether the model has learned useful patterns and whether it can make reliable predictions on unseen data.

For a classification problem such as Titanic survival prediction, evaluation metrics are used to compare the model's predicted outcomes with the actual survival outcomes.

The actual values are the recorded survival outcomes in the dataset, while the predicted values are the outcomes generated by the machine learning model.

## Accuracy

Accuracy measures the proportion of correct predictions made by the model out of all predictions.

The formula is:

Accuracy = (Correct Predictions) / (Total Predictions)

A higher accuracy value indicates that the model correctly predicted a larger percentage of observations.

However, accuracy alone may not always provide a complete picture, especially when classes are imbalanced.

For the Titanic dataset, accuracy represents the percentage of passengers for whom the model correctly predicted whether they survived or did not survive.


## Confusion Matrix

A confusion matrix is a table used to evaluate the performance of a classification model.

It compares the actual outcomes with the predicted outcomes.

The confusion matrix contains four values:

- **True Positive (TP):** The model correctly predicted a passenger survived.
- **True Negative (TN):** The model correctly predicted a passenger did not survive.
- **False Positive (FP):** The model predicted survival, but the passenger did not survive.
- **False Negative (FN):** The model predicted that a passenger did not survive, but the passenger actually survived.

The confusion matrix helps identify the types of errors made by the model.

## Precision

Precision measures how many of the passengers predicted as survivors actually survived.

Formula:

Precision = True Positives / (True Positives + False Positives)

A high precision value means that when the model predicts survival, it is usually correct.

## Recall

Recall measures how many of the actual survivors were correctly identified by the model.

Formula:

Recall = True Positives / (True Positives + False Negatives)

A high recall value means the model successfully identifies most passengers who survived.

## Determining Model Performance

A model should not be evaluated using only one metric. Different evaluation measures provide different information about model performance.

For the Titanic dataset:
- Accuracy shows the overall percentage of correct predictions.
- The confusion matrix shows the types of prediction errors.
- Precision shows how reliable survival predictions are.
- Recall shows how well the model identifies actual survivors.

A model performs well when it achieves good results on unseen testing data and maintains a balance between correctly identifying survivors and avoiding incorrect predictions.